server/src/main/scala/com/cloudera/livy/server/Main.scala

```diff
@@ -83,7 +85,8 @@ object Main extends Logging {
 
     livyConf.get(AUTH_TYPE) match {
       case authType @ KerberosAuthenticationHandler.TYPE =>
-        val principal = livyConf.get(KERBEROS_PRINCIPAL)
+        val principal = SecurityUtil.getServerPrincipal(livyConf.get(KERBEROS_PRINCIPAL),
+          InetAddress.getLocalHost.getCanonicalHostName)
```

Comment:
This doesn't really work. This is what I get on my machine:

```
scala> java.net.InetAddress.getLocalHost.getCanonicalHostName
res0: String = localhost
```

You need to use the `SERVER_HOST` configuration here, and have a fallback for when it's set to the wildcard address. In fact, because of how SPNEGO works, `SERVER_HOST` probably cannot be the wildcard address when security is enabled, since clients have to request a ticket that matches that address.

So, basically, I think this should be how this works:
- first check whether the principal contains `SecurityUtils.HOSTNAME_PATTERN`; if it doesn't, don't event try the options below, just use it.
- check `SERVER_HOST`, if it's the wildcard address, print a warning, and don't use it
- try `InetAddress.getLocalHost`; if `isLoopbackAddress()` is true, print a warning. Or maybe even error out.